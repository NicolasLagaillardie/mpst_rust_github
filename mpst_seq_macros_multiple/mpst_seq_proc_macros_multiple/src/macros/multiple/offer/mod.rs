pub mod create_offer_mpst_session_multi;
pub mod create_offer_type_multi;
