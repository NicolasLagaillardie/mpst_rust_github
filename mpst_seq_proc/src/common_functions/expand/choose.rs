use proc_macro2::{Span, TokenStream};
use quote::quote;
use std::convert::TryFrom;
use syn::Ident;

use crate::common_functions::maths::{
    diag_and_matrix, get_line_column_from_diag, get_tuple_matrix,
};

/// Expand choose methods
pub(crate) fn choose(
    all_roles: Vec<TokenStream>,
    sender: u64,
    meshedchannels_name: Ident,
    number_roles: u64,
) -> TokenStream {
    let (diag, matrix) = diag_and_matrix(number_roles);

    let sender_ident = if let Some(elt) = all_roles.get(usize::try_from(sender - 1).unwrap()) {
        Ident::new(&format!("Name{}", elt), Span::call_site())
    } else {
        panic!("Not enough arguments for sender_ident in expand_choose")
    };

    let sender_stack = if let Some(elt) = all_roles.get(usize::try_from(sender - 1).unwrap()) {
        Ident::new(&format!("Role{}toAll", elt), Span::call_site())
    } else {
        panic!("Not enough arguments for sender_stack in expand_choose")
    };

    let choose_session_types_struct: Vec<TokenStream> = (1..=((number_roles - 1) * number_roles))
        .map(|i| {
            let temp_ident = Ident::new(&format!("S{}", i), Span::call_site());
            quote! { #temp_ident : mpstthree::binary::struct_trait::session::Session , }
        })
        .collect();

    let choose_roles_struct: Vec<TokenStream> = (1..=(2 * number_roles))
        .map(|i| {
            let temp_ident = Ident::new(&format!("R{}", i), Span::call_site());
            quote! { #temp_ident : mpstthree::role::Role , }
        })
        .collect();

    let choose_sessions: Vec<TokenStream> = (1..=number_roles)
            .map(|j| {
                if sender != j {
                    let left_sessions: Vec<TokenStream> = (1..number_roles)
                        .map(|k| {

                            let (l, _, _) = get_tuple_matrix(&matrix, j, k);

                            if l == 0 {
                                panic!("Erratum choose_sessions / left_sessions j = {:?}", j)
                            };

                            let (_, _, m1) = if j > sender {
                                get_tuple_matrix(&matrix, sender, j - 1)
                            } else {
                                get_tuple_matrix(&matrix, sender, j)
                            };
                            let (_, _, m2) = get_tuple_matrix(&matrix, j, k);

                            let (_, _, m) = get_tuple_matrix(&matrix, j, k);

                            let temp_ident = Ident::new(
                                &format!("S{}", m),
                                Span::call_site(),
                            );

                            if l == j || m1 == m2 {
                                quote! { #temp_ident , }
                            } else {
                                quote! { <#temp_ident as mpstthree::binary::struct_trait::session::Session>::Dual , }
                            }
                        })
                        .collect();

                    let right_sessions: Vec<TokenStream> = (1..number_roles)
                        .map(|k| {

                            let (l, _, _) = get_tuple_matrix(&matrix, j, k);

                            if l == 0 {
                                panic!("Erratum choose_sessions / right_sessions j = {:?}", j)
                            };

                            let (_, _, m1) = if j > sender {
                                get_tuple_matrix(&matrix, sender, j - 1)
                            } else {
                                get_tuple_matrix(&matrix, sender, j)
                            };
                            let (_, _, m2) = get_tuple_matrix(&matrix, j, k);

                            let (_, _, m) = get_tuple_matrix(&matrix, j, k);

                            let diff = number_roles - 1;

                            let temp_ident = Ident::new(
                                &format!("S{}", diff * (diff + 1) / 2 + m),
                                Span::call_site(),
                            );

                            if l == j || m1 == m2 {
                                quote! { #temp_ident , }
                            } else {
                                quote! { <#temp_ident as mpstthree::binary::struct_trait::session::Session>::Dual , }
                            }
                        })
                        .collect();

                    let stack_left = if j > sender {
                        let temp_ident = Ident::new(
                            &format!("R{}", 2 * (j - 1) - 1),
                            Span::call_site(),
                        );
                        quote! { #temp_ident , }
                    } else {
                        let temp_ident = Ident::new(
                            &format!("R{}", 2 * (j - 1) + 1),
                            Span::call_site(),
                        );
                        quote! { #temp_ident , }
                    };

                    let stack_right = if j > sender {
                        let temp_ident = Ident::new(
                            &format!("R{}", 2 * (j - 1)),
                            Span::call_site(),
                        );
                        quote! { #temp_ident , }
                    } else {
                        let temp_ident = Ident::new(
                            &format!("R{}", 2 * (j - 1) + 2),
                            Span::call_site(),
                        );
                        quote! { #temp_ident , }
                    };

                    let receiver_ident =
                        if let Some(elt) = all_roles.get(usize::try_from(j - 1).unwrap()) {
                            Ident::new(&format!("Name{}", elt), Span::call_site())
                        } else {
                            panic!("Not enough arguments for receiver_ident in choose_sessions in expand_choose")
                        };

                    quote! {
                        mpstthree::binary::struct_trait::send::Send<
                            either::Either<
                                #meshedchannels_name<
                                    #(
                                        #left_sessions
                                    )*
                                    #stack_left
                                    #receiver_ident
                                >,
                                #meshedchannels_name<
                                    #(
                                        #right_sessions
                                    )*
                                    #stack_right
                                    #receiver_ident
                                >
                            >,
                            mpstthree::binary::struct_trait::end::End,
                        >,
                    }
                } else {
                    quote! {
                        // Empty
                    }
                }
            })
            .collect();

    let new_stack_sender_left =
        Ident::new(&format!("R{}", 2 * number_roles - 1), Span::call_site());
    let new_stack_sender_right = Ident::new(&format!("R{}", 2 * number_roles), Span::call_site());
    let new_stacks_sender = quote! { #new_stack_sender_left , #new_stack_sender_right };

    let choose_left_session: Vec<TokenStream> = (1..=number_roles)
            .filter_map(|j| {
                if j == sender {
                    None
                } else {
                    let (_, _, m) = if j > sender {
                        get_tuple_matrix(&matrix, sender, j - 1)
                    } else {
                        get_tuple_matrix(&matrix, sender, j)
                    };
                    let temp_ident =
                        Ident::new(&format!("S{}", m), Span::call_site());
                    Some(
                        quote! { <#temp_ident as mpstthree::binary::struct_trait::session::Session>::Dual, },
                    )
                }
            })
            .collect();

    let choose_right_session: Vec<TokenStream> = (1..=number_roles)
            .filter_map(|j| {
                if j == sender {
                    None
                } else {
                    let (_, _, m) = if j > sender {
                        get_tuple_matrix(&matrix, sender, j - 1)
                    } else {
                        get_tuple_matrix(&matrix, sender, j)
                    };
                    let diff = number_roles - 1;
                    let temp_ident = Ident::new(
                        &format!("S{}", diff * (diff + 1) / 2 + m),
                        Span::call_site(),
                    );
                    Some(
                        quote! { <#temp_ident as mpstthree::binary::struct_trait::session::Session>::Dual, },
                    )
                }
            })
            .collect();
    let choose_left_channels: Vec<TokenStream> = (1..=((number_roles - 1) * number_roles / 2))
        .map(|j| {
            let (line, column) = get_line_column_from_diag(&diag, j);

            let first_channel = if sender != line {
                Ident::new(&format!("channel_{}_{}", line, column), Span::call_site())
            } else {
                Ident::new(&format!("channel_{}_{}", column, line), Span::call_site())
            };

            let second_channel = if sender != line {
                Ident::new(&format!("channel_{}_{}", column, line), Span::call_site())
            } else {
                Ident::new(&format!("channel_{}_{}", line, column), Span::call_site())
            };

            let temp_session = Ident::new(&format!("S{}", j), Span::call_site());

            quote! { let ( #first_channel , #second_channel ) =
            <#temp_session as mpstthree::binary::struct_trait::session::Session>::new() ; }
        })
        .collect();

    let choose_right_channels: Vec<TokenStream> = (1..=((number_roles - 1) * number_roles / 2))
        .map(|j| {
            let (line, column) = get_line_column_from_diag(&diag, j);
            let diff = number_roles - 1;

            let first_channel = if sender != line {
                Ident::new(&format!("channel_{}_{}", line, column), Span::call_site())
            } else {
                Ident::new(&format!("channel_{}_{}", column, line), Span::call_site())
            };

            let second_channel = if sender != line {
                Ident::new(&format!("channel_{}_{}", column, line), Span::call_site())
            } else {
                Ident::new(&format!("channel_{}_{}", line, column), Span::call_site())
            };

            let temp_session = Ident::new(
                &format!("S{}", diff * (diff + 1) / 2 + j),
                Span::call_site(),
            );

            quote! { let ( #first_channel , #second_channel ) = < #temp_session as mpstthree::binary::struct_trait::session::Session >::new() ; }
        })
        .collect();

    let new_stacks_receivers_left: Vec<TokenStream> = (1..number_roles)
        .map(|j| {
            let temp_stack = Ident::new(&format!("stack_{}", j), Span::call_site());
            let temp_role = Ident::new(&format!("R{}", 2 * (j - 1) + 1), Span::call_site());
            quote! { let (#temp_stack, _) = <#temp_role as mpstthree::role::Role>::new(); }
        })
        .collect();

    let new_stacks_receivers_right: Vec<TokenStream> = (1..number_roles)
        .map(|j| {
            let temp_stack = Ident::new(&format!("stack_{}", j), Span::call_site());
            let temp_role = Ident::new(&format!("R{}", 2 * (j - 1) + 2), Span::call_site());
            quote! { let (#temp_stack, _) = <#temp_role as mpstthree::role::Role>::new(); }
        })
        .collect();

    let new_names: Vec<TokenStream> = (1..=number_roles)
        .map(|j| {
            if sender != j {
                let receiver_ident = if let Some(elt) =
                    all_roles.get(usize::try_from(j - 1).unwrap())
                {
                    Ident::new(&format!("Name{}", elt), Span::call_site())
                } else {
                    panic!("Not enough arguments for receiver_ident in new_names in expand_choose")
                };

                let new_name = if let Some(elt) = all_roles.get(usize::try_from(j - 1).unwrap()) {
                    Ident::new(&format!("name_{}", elt), Span::call_site())
                } else {
                    panic!("Not enough arguments for new_name in new_names in expand_choose")
                };

                quote! {
                    let (#new_name, _) = < #receiver_ident as mpstthree::name::Name >::new();
                }
            } else {
                quote! {}
            }
        })
        .collect();

    let new_meshedchannels_receivers: Vec<TokenStream> = (1..=number_roles)
            .map(|j| {
                if sender != j {
                    let new_sessions_receiver: Vec<TokenStream> = (1..number_roles)
                        .map(|k| {
                            let new_session_receiver = Ident::new(
                                &format!("session{}", k),
                                Span::call_site(),
                            );
                            let new_channel_receiver = if j > k {
                                Ident::new(
                                    &format!("channel_{}_{}", j, k),
                                    Span::call_site(),
                                )
                            } else {
                                Ident::new(
                                    &format!("channel_{}_{}", j, k + 1),
                                    Span::call_site(),
                                )
                            };

                            quote! { #new_session_receiver : #new_channel_receiver , }
                        })
                        .collect();

                    let new_choice_receiver = if j > sender
                    {
                        Ident::new(&format!("choice_{}", j - 1), Span::call_site())
                    } else {
                        Ident::new(&format!("choice_{}", j), Span::call_site())
                    };

                    let new_stack_receiver = if j > sender
                    {
                        Ident::new(&format!("stack_{}", j - 1), Span::call_site())
                    } else {
                        Ident::new(&format!("stack_{}", j), Span::call_site())
                    };

                    let new_name_receiver = if let Some(elt) =
                        all_roles.get(usize::try_from(j - 1).unwrap())
                    {
                        Ident::new(&format!("name_{}", elt), Span::call_site())
                    } else {
                        panic!("Not enough arguments for new_name_receiver in new_meshedchannels_receivers in expand_choose")
                    };

                    quote! {
                        let #new_choice_receiver = #meshedchannels_name {
                            #(
                                #new_sessions_receiver
                            )*
                            stack: #new_stack_receiver,
                            name: #new_name_receiver,
                        };
                    }
                } else {
                    quote! {
                        // Empty
                    }
                }
            })
            .collect();

    let new_sessions_sender_left: Vec<TokenStream> = (1..number_roles)
        .map(|j| {
            let new_session_sender =
                Ident::new(&format!("new_session_{}", j - 1), Span::call_site());

            let new_choice_sender = Ident::new(&format!("choice_{}", j), Span::call_site());

            let session_sender = Ident::new(&format!("session{}", j), Span::call_site());

            quote! {
                let #new_session_sender = mpstthree::binary::send::send(
                    either::Either::Left(#new_choice_sender),
                    self.#session_sender
                );
            }
        })
        .collect();

    let new_sessions_sender_right: Vec<TokenStream> = (1..number_roles)
        .map(|j| {
            let new_session_sender =
                Ident::new(&format!("new_session_{}", j - 1), Span::call_site());

            let new_choice_sender = Ident::new(&format!("choice_{}", j), Span::call_site());

            let session_sender = Ident::new(&format!("session{}", j), Span::call_site());

            quote! {
                let #new_session_sender = mpstthree::binary::send::send(
                    either::Either::Right(#new_choice_sender),
                    self.#session_sender
                );
            }
        })
        .collect();

    let old_meshedchannels_sender: Vec<TokenStream> = (1..number_roles)
        .map(|j| {
            let new_session_sender =
                Ident::new(&format!("new_session_{}", j - 1), Span::call_site());

            let session_sender = Ident::new(&format!("session{}", j), Span::call_site());

            quote! {
                #session_sender : #new_session_sender ,
            }
        })
        .collect();

    let new_meshedchannels_sender: Vec<TokenStream> = (1..=number_roles)
        .map(|j| {
            if sender != j {
                let new_choice_sender = if j < sender {
                    Ident::new(&format!("session{}", j), Span::call_site())
                } else {
                    Ident::new(&format!("session{}", j - 1), Span::call_site())
                };

                let new_channel_sender =
                    Ident::new(&format!("channel_{}_{}", sender, j), Span::call_site());

                quote! {
                    #new_choice_sender : #new_channel_sender,
                }
            } else {
                quote! {
                    // Empty
                }
            }
        })
        .collect();

    let new_stack_sender = Ident::new(&format!("stack_{}", number_roles), Span::call_site());

    let new_name_sender = Ident::new(&format!("name_{}", number_roles), Span::call_site());

    quote! {
        impl<
            #(
                #choose_session_types_struct
            )*
            #(
                #choose_roles_struct
            )*
        >
            #meshedchannels_name<
                #(
                    #choose_sessions
                )*
                #sender_stack<
                    #new_stacks_sender
                >,
                #sender_ident,
            >
        {
            pub fn choose_left(self) -> #meshedchannels_name<
                #(
                    #choose_left_session
                )*
                #new_stack_sender_left ,
                #sender_ident
            >
            {
                #(
                    #choose_left_channels
                )*

                #(
                    #new_stacks_receivers_left
                )*

                let (#new_stack_sender, _) = <#new_stack_sender_left as mpstthree::role::Role>::new();

                #(
                    #new_names
                )*

                let (#new_name_sender, _) = < #sender_ident as mpstthree::name::Name >::new();

                #(
                    #new_meshedchannels_receivers
                )*

                #(
                    #new_sessions_sender_left
                )*

                let s = #meshedchannels_name {
                    #(
                        #old_meshedchannels_sender
                    )*
                    stack: self.stack,
                    name: self.name,
                };

                mpstthree::binary::cancel::cancel(s);

                #meshedchannels_name {
                    #(
                        #new_meshedchannels_sender
                    )*
                    stack: #new_stack_sender,
                    name: #new_name_sender,
                }
            }
            pub fn choose_right(self) -> #meshedchannels_name<
                #(
                    #choose_right_session
                )*
                #new_stack_sender_right ,
                #sender_ident
            >
            {
                #(
                    #choose_right_channels
                )*

                #(
                    #new_stacks_receivers_right
                )*

                let (#new_stack_sender, _) = <#new_stack_sender_right as mpstthree::role::Role>::new();

                #(
                    #new_names
                )*

                let (#new_name_sender, _) = < #sender_ident as mpstthree::name::Name >::new();

                #(
                    #new_meshedchannels_receivers
                )*

                #(
                    #new_sessions_sender_right
                )*

                let s = #meshedchannels_name {
                    #(
                        #old_meshedchannels_sender
                    )*
                    stack: self.stack,
                    name: self.name,
                };

                mpstthree::binary::cancel::cancel(s);

                #meshedchannels_name {
                    #(
                        #new_meshedchannels_sender
                    )*
                    stack: #new_stack_sender,
                    name: #new_name_sender,
                }
            }
        }
    }
}

/// Expand choose methods for interleaved sessions
pub(crate) fn choose_mpst_create_multi_to_all(
    meshedchannels_name: Ident,
    all_roles: Vec<TokenStream>,
    number_roles: u64,
) -> TokenStream {
    let choose_mpst_create_multi_to_all: Vec<TokenStream> = (1..=number_roles)
        .map(|sender| {

            let name_macro = if let Some(elt) =
                all_roles.get(usize::try_from(sender - 1).unwrap())
            {
                Ident::new(
                    &format!("choose_mpst_{}_to_all", elt).to_lowercase(),
                    Span::call_site(),
                )
            } else {
                panic!("Not enough arguments for name in expand_choose_mpst_create_multi_to_all")
            };

            let sender_name = if let Some(elt) =
                all_roles.get(usize::try_from(sender - 1).unwrap())
            {
                Ident::new(
                    &format!("Name{}", elt),
                    Span::call_site(),
                )
            } else {
                panic!("Not enough arguments for sender_name in expand_choose_mpst_create_multi_to_all")
            };

            let receivers: Vec<Ident> = (1..=number_roles)
                .filter_map(|receiver| {
                    if sender != receiver {
                        Some(
                            if let Some(elt) =
                                all_roles.get(usize::try_from(receiver - 1).unwrap())
                            {
                                Ident::new(
                                    &format!("Name{}", elt),
                                    Span::call_site(),
                                )
                            } else {
                                panic!("Not enough arguments for receivers in expand_choose_mpst_create_multi_to_all")
                            }
                        )
                    } else {
                        None
                    }
                })
                .collect();

            quote! {
                mpstthree::choose_mpst_create_multi_to_all!(
                    #name_macro ,
                    #( #receivers , )* =>
                    #sender_name ,
                    #meshedchannels_name ,
                    #sender
                );
            }
        })
        .collect();

    quote! {
        #( #choose_mpst_create_multi_to_all )*
    }
}
