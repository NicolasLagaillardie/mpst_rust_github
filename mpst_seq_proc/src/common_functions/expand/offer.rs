use proc_macro2::{Span, TokenStream};
use quote::quote;
use std::convert::TryFrom;
use syn::Ident;

/// Expand offer methods
pub(crate) fn offer(
    all_roles: &[TokenStream],
    sender: u64,
    receiver: u64,
    meshedchannels_name: &Ident,
    number_roles: u64,
) -> TokenStream {
    let sender_ident = if let Some(elt) = all_roles.get(usize::try_from(sender - 1).unwrap()) {
        Ident::new(&format!("RoleAllto{}", elt), Span::call_site())
    } else {
        panic!("Not enough arguments for sender_ident in expand_offer")
    };

    let receiver_ident = if let Some(elt) = all_roles.get(usize::try_from(receiver - 1).unwrap()) {
        Ident::new(&format!("Name{}", elt), Span::call_site())
    } else {
        panic!("Not enough arguments for receiver_ident in expand_offer")
    };

    let offer_session_types_struct: Vec<TokenStream> = (1..(2 * number_roles - 1))
        .map(|i| {
            let temp_ident = Ident::new(&format!("S{}", i), Span::call_site());
            quote! { #temp_ident : mpstthree::binary::struct_trait::session::Session , }
        })
        .collect();

    let left_sessions: Vec<TokenStream> = (1..number_roles)
        .map(|i| {
            let temp_ident = Ident::new(&format!("S{}", i), Span::call_site());
            quote! { #temp_ident , }
        })
        .collect();

    let right_sessions: Vec<TokenStream> = (number_roles..(2 * number_roles - 1))
        .map(|i| {
            let temp_ident = Ident::new(&format!("S{}", i), Span::call_site());
            quote! { #temp_ident , }
        })
        .collect();

    let offer_sessions: Vec<TokenStream> = (1..number_roles)
        .map(|k| {
            let cond = if k >= receiver { sender - 1 } else { sender };
            if k == cond {
                quote! {
                    mpstthree::binary::struct_trait::recv::Recv<
                        either::Either<
                            #meshedchannels_name<
                                #( #left_sessions )*
                                R1,
                                #receiver_ident
                            >,
                            #meshedchannels_name<
                                #( #right_sessions )*
                                R2,
                                #receiver_ident
                            >
                        >,
                        mpstthree::binary::struct_trait::end::End
                    >,
                }
            } else {
                quote! { mpstthree::binary::struct_trait::end::End, }
            }
        })
        .collect();

    quote! {
        impl<
            'a,
            #( #offer_session_types_struct )*
            R1: mpstthree::role::Role,
            R2: mpstthree::role::Role,
        >
            #meshedchannels_name<
                #( #offer_sessions )*
                #sender_ident<mpstthree::role::end::RoleEnd, mpstthree::role::end::RoleEnd>,
                #receiver_ident,
            >
        {
            pub fn offer<F, G, U>(self, f: F, g: G) -> Result<U, Box<dyn std::error::Error + 'a>>
            where
                F: FnOnce(
                    #meshedchannels_name<
                        #( #left_sessions )*
                        R1,
                        #receiver_ident,
                    >,
                ) -> Result<U, Box<dyn std::error::Error + 'a>>,
                G: FnOnce(
                    #meshedchannels_name<
                        #( #right_sessions )*
                        R2,
                        #receiver_ident,
                    >,
                ) -> Result<U, Box<dyn std::error::Error + 'a>>,
            {
                let (e, s) = self.recv_from_all()?;
                s.cancel();
                e.either(f, g)
            }
        }
    }
}

/// Expand offer methods
pub(crate) fn offer_timed(
    all_roles: &[TokenStream],
    sender: u64,
    receiver: u64,
    meshedchannels_name: &Ident,
    number_roles: u64,
) -> TokenStream {
    let sender_ident = if let Some(elt) = all_roles.get(usize::try_from(sender - 1).unwrap()) {
        Ident::new(&format!("RoleAllto{}", elt), Span::call_site())
    } else {
        panic!("Not enough arguments for sender_ident in expand_offer")
    };

    let receiver_ident = if let Some(elt) = all_roles.get(usize::try_from(receiver - 1).unwrap()) {
        Ident::new(&format!("Name{}", elt), Span::call_site())
    } else {
        panic!("Not enough arguments for receiver_ident in expand_offer")
    };

    let offer_session_types_struct: Vec<TokenStream> = (1..(2 * number_roles - 1))
        .map(|i| {
            let temp_ident = Ident::new(&format!("S{}", i), Span::call_site());
            quote! { #temp_ident : mpstthree::binary::struct_trait::session::Session , }
        })
        .collect();

    let left_sessions: Vec<TokenStream> = (1..number_roles)
        .map(|i| {
            let temp_ident = Ident::new(&format!("S{}", i), Span::call_site());
            quote! { #temp_ident , }
        })
        .collect();

    let right_sessions: Vec<TokenStream> = (number_roles..(2 * number_roles - 1))
        .map(|i| {
            let temp_ident = Ident::new(&format!("S{}", i), Span::call_site());
            quote! { #temp_ident , }
        })
        .collect();

    let offer_sessions: Vec<TokenStream> = (1..number_roles)
        .map(|k| {
            let cond = if k >= receiver { sender - 1 } else { sender };
            if k == cond {
                quote! {
                    mpstthree::binary_timed::struct_trait::recv::RecvTimed<
                        either::Either<
                            #meshedchannels_name<
                                #( #left_sessions )*
                                R1,
                                #receiver_ident
                            >,
                            #meshedchannels_name<
                                #( #right_sessions )*
                                R2,
                                #receiver_ident
                            >
                        >,
                        mpstthree::binary::struct_trait::end::End,
                        CLOCK,
                        START,
                        INCLUDE_START,
                        END,
                        INCLUDE_END,
                        RESET
                    >,
                }
            } else {
                quote! { mpstthree::binary::struct_trait::end::End, }
            }
        })
        .collect();

    quote! {
        impl<
            'a,
            #( #offer_session_types_struct )*
            R1: mpstthree::role::Role,
            R2: mpstthree::role::Role,
            const CLOCK: char,
            const START: i128,
            const INCLUDE_START: bool,
            const END: i128,
            const INCLUDE_END: bool,
            const RESET: bool,
        >
            #meshedchannels_name<
                #( #offer_sessions )*
                #sender_ident<mpstthree::role::end::RoleEnd, mpstthree::role::end::RoleEnd>,
                #receiver_ident,
            >
        {
            pub fn offer<F, G, U>(
                self,
                all_clocks: &mut std::collections::HashMap<char, std::time::Instant>,
                f: F,
                g: G
            ) -> Result<U, Box<dyn std::error::Error + 'a>>
            where
                F: FnOnce(
                    &mut std::collections::HashMap<char, std::time::Instant>,
                    #meshedchannels_name<
                        #( #left_sessions )*
                        R1,
                        #receiver_ident,
                    >,
                ) -> Result<U, Box<dyn std::error::Error + 'a>>,
                G: FnOnce(
                    &mut std::collections::HashMap<char, std::time::Instant>,
                    #meshedchannels_name<
                        #( #right_sessions )*
                        R2,
                        #receiver_ident,
                    >,
                ) -> Result<U, Box<dyn std::error::Error + 'a>>,
            {
                let (e, s) = self.recv_from_all(all_clocks)?;
                s.cancel();
                e.either_with(all_clocks, f, g)
            }
        }
    }
}
