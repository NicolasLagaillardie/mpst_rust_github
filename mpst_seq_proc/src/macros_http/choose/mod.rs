pub mod choose_mpst_multi_http_to_all;
