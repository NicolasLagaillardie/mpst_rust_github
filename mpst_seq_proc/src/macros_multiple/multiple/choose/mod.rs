pub(crate) mod choose_atmp_mpst_create_multi_to_all;
pub(crate) mod choose_mpst_create_multi_to_all;
pub(crate) mod choose_mpst_multi_cancel_to_all;
pub(crate) mod create_choose_mpst_session_multi_left;
pub(crate) mod create_choose_mpst_session_multi_right;
pub(crate) mod create_choose_type_multi;
pub(crate) mod create_fn_choose_mpst_cancel_multi_to_all_bundle;
pub(crate) mod create_fn_choose_mpst_multi_to_all_bundle;
