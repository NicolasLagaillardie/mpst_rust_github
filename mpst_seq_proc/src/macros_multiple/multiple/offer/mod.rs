pub mod create_offer_type_multi;
