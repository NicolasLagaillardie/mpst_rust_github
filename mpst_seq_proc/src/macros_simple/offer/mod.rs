pub mod create_offer_mpst_session_multi;
