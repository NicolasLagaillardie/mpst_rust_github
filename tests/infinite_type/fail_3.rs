use mpstthree::binary::struct_trait::{end::End, recv::Recv, send::Send, session::Session};
use mpstthree::binary::*;
use mpstthree::choose;
use mpstthree::offer;

use rand::{thread_rng, Rng};

use std::boxed::Box;
use std::error::Error;

enum SumOp {
    More(Send<i32, Recv<i32, Send<i32, Recv<i32, Recv<Self, End>>>>>),
    Done(Send<i32, End>),
}
type NiceSumServer = Recv<SumOp, End>;
type NiceSumClient = <NiceSumServer as Session>::Dual;

fn nice_sum_server(s: NiceSumServer) -> Result<(), Box<dyn Error>> {
    nice_sum_server_accum(s, 0)
}

fn nice_sum_server_accum(s: NiceSumServer, x: i32) -> Result<(), Box<dyn Error>> {
    offer!(s, {
        SumOp::More(s) => {
            let s = send::send(0, s);
            let (y, s) = recv::recv(s)?;
            let s = send::send(0, s);
            let (y, s) = recv::recv(s)?;
            nice_sum_server_accum(s, x.wrapping_add(y))
        },
        SumOp::Done(s) => {
            let s = send::send(x, s);
            close::close(s)
        },
    })?;
    Ok(())
}

fn nice_sum_client_accum(s: NiceSumClient, mut xs: Vec<i32>) -> Result<i32, Box<dyn Error>> {
    match xs.pop() {
        Option::Some(x) => {
            let s = choose!(SumOp::More, s);
            let (_, s) = recv::recv(s)?;
            let s = send::send(x, s);
            let (_, s) = recv::recv(s)?;
            let s = send::send(x, s);
            nice_sum_client_accum(s, xs)
        }
        Option::None => {
            let s = choose!(SumOp::Done, s);
            let (sum, s) = recv::recv(s)?;
            close::close(s)?;
            Ok(sum)
        }
    }
}

fn main() {
    // Pick some random numbers.
    let mut rng = thread_rng();
    let xs: Vec<i32> = (1..100).map(|_| rng.gen()).collect();
    let sum1: i32 = xs.iter().fold(0, |sum, &x| sum.wrapping_add(x));

    let (other_thread, s) = fork::fork_with_thread_id(nice_sum_server);

    assert!({
        let sum2 = nice_sum_client_accum(s, xs)?;
        assert_eq!(sum1, sum2);
        Ok(())
    }
    .is_ok());

    assert!(other_thread.join().is_ok());
}
