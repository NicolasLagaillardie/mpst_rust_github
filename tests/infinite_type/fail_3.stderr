error[E0275]: overflow evaluating the requirement `mpstthree::binary::struct_trait::send::Send<i32, mpstthree::binary::struct_trait::send::Send<SumOp, mpstthree::binary::struct_trait::end::End>>: Session`
  --> tests/infinite_type/fail_3.rs:12:10
   |
12 |     More(Send<i32, Recv<i32, Send<i32, Recv<i32, Recv<Self, End>>>>>),
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
  ::: src/binary/struct_trait/send.rs:20:14
   |
20 |     S::Dual: Session,
   |              ------- required by this bound in `mpstthree::binary::struct_trait::send::Send`

error[E0275]: overflow evaluating the requirement `mpstthree::binary::struct_trait::send::Send<i32, Recv<i32, mpstthree::binary::struct_trait::send::Send<i32, Recv<i32, Recv<SumOp, mpstthree::binary::struct_trait::end::End>>>>>: std::marker::Send`
  --> tests/infinite_type/fail_3.rs:18:23
   |
18 | fn nice_sum_server(s: NiceSumServer) -> Result<(), Box<dyn Error>> {
   |                       ^^^^^^^^^^^^^
   |
  ::: src/binary/struct_trait/recv.rs:18:8
   |
18 |     T: marker::Send,
   |        ------------ required by this bound in `Recv`
   |
note: required because it appears within the type `SumOp`
  --> tests/infinite_type/fail_3.rs:11:6
   |
11 | enum SumOp {
   |      ^^^^^

error[E0275]: overflow evaluating the requirement `mpstthree::binary::struct_trait::send::Send<i32, Recv<i32, mpstthree::binary::struct_trait::send::Send<i32, Recv<i32, Recv<SumOp, mpstthree::binary::struct_trait::end::End>>>>>: std::marker::Send`
  --> tests/infinite_type/fail_3.rs:22:29
   |
22 | fn nice_sum_server_accum(s: NiceSumServer, x: i32) -> Result<(), Box<dyn Error>> {
   |                             ^^^^^^^^^^^^^
   |
  ::: src/binary/struct_trait/recv.rs:18:8
   |
18 |     T: marker::Send,
   |        ------------ required by this bound in `Recv`
   |
note: required because it appears within the type `SumOp`
  --> tests/infinite_type/fail_3.rs:11:6
   |
11 | enum SumOp {
   |      ^^^^^

error[E0275]: overflow evaluating the requirement `Recv<SumOp, mpstthree::binary::struct_trait::end::End>: Session`
  --> tests/infinite_type/fail_3.rs:39:4
   |
39 | fn nice_sum_client_accum(s: NiceSumClient, mut xs: Vec<i32>) -> Result<i32, Box<dyn Error>> {
   |    ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: required because of the requirements on the impl of `Session` for `Recv<i32, Recv<SumOp, mpstthree::binary::struct_trait::end::End>>`
   = note: required because it appears within the type `mpstthree::binary::struct_trait::send::Send<i32, Recv<i32, mpstthree::binary::struct_trait::send::Send<i32, Recv<i32, Recv<SumOp, mpstthree::binary::struct_trait::end::End>>>>>`
note: required because it appears within the type `SumOp`
  --> tests/infinite_type/fail_3.rs:11:6
   |
11 | enum SumOp {
   |      ^^^^^
   = note: required because of the requirements on the impl of `Session` for `Recv<SumOp, mpstthree::binary::struct_trait::end::End>`
